### PR TITLE
Add a write_char method to std::fmt::Formatter.

### PR DESCRIPTION
**Pull request by SimonSapin**
This is the logical next step after #24661, but I’m less sure about this one.

r? @alexcrichton

### CODE REVIEW

**Comment by alexcrichton** [comment]
Hm can you clarify why you'd want this as an inherent method? It seems like a similar win is gained from implementing `fmt::Write`. The main reason `fmt::Write` isn't leveraged for `write_fmt` is that it would require an import of a somewhat obscure trait in lots of code, but I think that `write_char` is pretty rare by comparison.

**Comment by SimonSapin** [comment]
I only made it an inherent method because `write_str` was as well, and this seemed like the smaller change. Note that `Formatter` currently does not implement any trait.

I don’t have a strong opinion on how to do this, or if it should be done at all.

**Comment by alexcrichton** [comment]
I think it would be safe to provide an implementation of the trait for `Formatter`, but I think we may want to hold off adding a method for now.

**Comment by alexcrichton** [comment]
@SimonSapin could you try adding this method through an implementation of the `Write` trait on the `Formatter` struct directly? I think it should have the implementation anyway and seems like a conservative way to expose the functionality.

**Comment by SimonSapin** [comment]
Done. (Now the trait has to be in scope to use that method, of course.)

File: src/libcore/fmt/mod.rs [hunk 1]
```diff
@@ -892,6 +892,20 @@ impl<'a> Formatter<'a> {
     }
 }
 
+impl<'a> Write for Formatter<'a> {
```
Comment (alexcrichton) [inline src/libcore/fmt/mod.rs]: Could you also tag this with `#[stable(since = "1.2.0", ...)]`?

**Comment by alexcrichton** [comment]
r=me with a stability tag, thanks @SimonSapin!

File: src/libcore/fmt/mod.rs [hunk 2]
```diff
@@ -965,10 +980,7 @@ impl Debug for char {
 #[stable(feature = "rust1", since = "1.0.0")]
 impl Display for char {
     fn fmt(&self, f: &mut Formatter) -> Result {
-        let mut utf8 = [0; 4];
-        let amt = self.encode_utf8(&mut utf8).unwrap_or(0);
-        let s: &str = unsafe { mem::transmute(&utf8[..amt]) };
-        Display::fmt(s, f)
```
Comment (BigBoneDaddy) [inline src/libcore/fmt/mod.rs]: If Display::fmt() not called is formatting lost?

Comment (SimonSapin) [inline src/libcore/fmt/mod.rs]: Ah, yes, my mistake. Feel free to reverse this chunk if there isn’t a better solution.

**Pull request merged**